atc/metric/emitter/newrelic.go

```diff
@@ -177,10 +221,139 @@ func (emitter *NewRelicEmitter) Emit(logger lager.Logger, event metric.Event) {
+		batchPayload := emitter.flush()
+		if batchPayload != nil && len(batchPayload) > 0 {
+			emitter.updateLastEmitTime()
+			go emitter.emitBatch(logger, batchPayload)
+		}
+	}
+}
+
+func (emitter *batchEmitter) Emit(logger lager.Logger, payload singlePayload) {
+	// enqueue the data
+	// check the time since last time we emitted the data, if it is great than the threshold, emit right now
+	payloadToSend, err := emitter.enqueue(payload)
+	if err != nil {
+		logger.Error(err.Action, err.Error)
+		return
+	}
+	if payloadToSend != nil {
+		emitter.updateLastEmitTime()
+		go emitter.emitBatch(logger, payloadToSend)
+	}
+
+	emitter.Flush(logger)
+}
+
+// enqueue new payload
+// if enqueue the new payload would exceed the max limit (1Mb)
+// return everything that's enqueued and enqueue the new payload
+// lock and unlock should be applied.
+func (emitter *batchEmitter) enqueue(newPayload singlePayload) (batchPayload, *loggableError) {
+	var (
+		tempBuff []byte
+		buff     batchPayload
+	)
+	newPayloadData, err := json.Marshal(newPayload)
+	if err != nil {
+		return nil, newLoggableError("failed-to-serialize-new-payload", err)
 	}
 
-	if len(payload) > 0 {
-		emitter.emitPayload(logger, payload)
+	if newPayloadData == nil || len(newPayloadData) == 0 {
+		return nil, newLoggableError("empty-new-payload", err)
 	}
+
+	if len(emitter.emitBuffer) != 0 {
+		tempBuff = append(emitter.emitBuffer, ',')
+	}
+	tempBuff = append(tempBuff, newPayloadData...)
+
+	payloadSize, sizeError := emitter.bufferSize(tempBuff)
+	if sizeError != nil {
+		return nil, sizeError
+	}
+
+	// when combined new payload exceeds 1MB
+	if payloadSize > newrelicPayloadMaxSize-2 { // reduce '[' and ']' for the json payload
+		buff = formatBatchEmitBuffer(emitter.emitBuffer)
+		emitter.emitBuffer = newPayloadData
+	} else {
+		emitter.emitBuffer = tempBuff
+	}
+	return buff, nil
+}
+
+func (emitter *batchEmitter) bufferSize(buff []byte) (int, *loggableError) {
+	payloadSize := len(buff)
+	// for better performance, we only check compressed data size when the data in the buffer is larger than 1MB
+	if emitter.compression && payloadSize >= newrelicPayloadMaxSize-2 { // reduce size for '[' and ']'
+		zippedBuffer, err := gZipBuffer(buff)
+		if err != nil {
+			return 0, newLoggableError("failed-to-gzip-payload", err)
+		}
+
+		zippedData, err := ioutil.ReadAll(zippedBuffer)
+		if err != nil {
+			return 0, newLoggableError("failed-to-use-gzip-buffer-payload", err)
+		}
+		payloadSize = len(zippedData)
+	}
+	return payloadSize, nil
+}
+
+func (emitter *batchEmitter) flush() batchPayload {
+	var buff batchPayload
+
+	if emitter.emitBuffer != nil {
+		buff = formatBatchEmitBuffer(emitter.emitBuffer)
+		emitter.emitBuffer = nil
+	}
+	return buff
+}
+
+func (emitter *batchEmitter) updateLastEmitTime() {
+	emitter.lastEmitTime = time.Now()
+}
+
+func (emitter *batchEmitter) flushThreshold() bool {
+	return time.Now().Sub(emitter.lastEmitTime) > emitter.flushInterval
```

Comment:
I was curious to see whether "time elapsed" calculations would be faster / slower than "ticker based" interval checking, and it turns out that "ticker based" seems to be faster lol

https://gist.github.com/cirocosta/9b19380ae14b1c2bf67438bb142b618a

BenchmarkTimeNow-12     15730425                73.0 ns/op
BenchmarkTicker-12      223836266                5.31 ns/op

I don't think we necessarily need to change this, just an interesting fact 😁